airbyte-integrations/connectors/destination-snowflake-cortex/destination_snowflake_cortex/indexer.py

```diff
@@ -0,0 +1,178 @@
+            stream.stream.json_schema["properties"][PAGE_CONTENT_COLUMN] = {"type": "string"}
+            stream.stream.json_schema["properties"][METADATA_COLUMN] = {"type": "object"}
+            stream.stream.json_schema["properties"][EMBEDDING_COLUMN] = {"type": "vector_array"}
+            # set primary key only if there are existing primary keys
+            if stream.primary_key:
+                stream.primary_key = [[DOCUMENT_ID_COLUMN]]
+        return updated_catalog
+
+    def _get_primary_keys(self, stream_name: str) -> Optional[str]:
+        for stream in self.catalog.streams:
+            if stream.stream.name == stream_name:
+                return stream.primary_key
+        return None
+
+    def _get_record_primary_key(self, record: AirbyteMessage) -> Optional[str]:
+        """Create primary key for the record by appending the primary keys."""
+        stream_name = record.record.stream
+        primary_keys = self._get_primary_keys(stream_name)
+
+        if not primary_keys:
+            return None
+
+        primary_key = []
+        for key in primary_keys:
+            try:
+                primary_key.append(str(dpath.util.get(record.record.data, key)))
+            except KeyError:
+                primary_key.append("__not_found__")
+        # return a stringified version of all primary keys
+        stringified_primary_key = "_".join(primary_key)
+        return stringified_primary_key
+
+    def _create_document_id(self, record: AirbyteMessage) -> str:
+        """Create document id based on the primary key values. Returns a random uuid if no primary key is found"""
+        stream_name = record.record.stream
+        primary_key = self._get_record_primary_key(record)
+        if primary_key is not None:
+            return f"Stream_{stream_name}_Key_{primary_key}"
+        return str(uuid.uuid4().int)
+
+    def _create_state_message(self, stream, namespace, data: dict[str, Any]) -> AirbyteMessage:
+        """Create a state message for the stream"""
+        stream = AirbyteStreamState(stream_descriptor=StreamDescriptor(name=stream, namespace=namespace))
+        return AirbyteMessage(
+            type=Type.STATE,
+            state=AirbyteStateMessage(type=AirbyteStateType.STREAM, stream=stream, data=data),
+        )
+
+    def get_write_strategy(self, stream_name: str) -> WriteStrategy:
+        for stream in self.catalog.streams:
+            if stream.stream.name == stream_name:
+                if stream.destination_sync_mode == DestinationSyncMode.overwrite:
+                    return WriteStrategy.REPLACE
+                if stream.destination_sync_mode == DestinationSyncMode.append:
+                    return WriteStrategy.APPEND
+                if stream.destination_sync_mode == DestinationSyncMode.append_dedup:
+                    return WriteStrategy.MERGE
+        return WriteStrategy.AUTO
+
+    def index(self, document_chunks, namespace, stream):
```

Comment:
Ditto above. Can we add type hints for `document_chunks`, `namespace`, and `stream`?

Reply:
done.